# from python to numpy

In [4]:
import numpy as np

In [7]:
import random

In [16]:
class RandomWalker:
    def __init__(self):
        self.position=0
    def walk(self,n):
        for i in range(n):
            yield self.position
            self.position+=2*random.choice([0,1])-1

In [17]:
walker=RandomWalker()

In [23]:
%%timeit
walk=[position for position in walker.walk(1000)]

761 µs ± 8.31 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


##### procedural approach

In [24]:
def randomwalk(n):
    position=0
    walk=[position]
    for i in range(n):
        walk.append(2*random.choice([0,1])-1)
    return walk
        

In [28]:

%%timeit
randomwalk(1000)


648 µs ± 5.63 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


##### using vectorisation approach

In [34]:
from itertools import accumulate 
import random
def random_vector_walker(n):
    steps=random.choices([-1,1],k=n)
    walk=[0]+list(accumulate(steps))
    return walk

    

In [36]:
%%timeit
walk=random_vector_walker(1000)

177 µs ± 6.54 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [37]:
import numpy as np

In [40]:
def walk_fast(n):
    steps=np.random.choice([1,-1],n)
    return np.cumsum(n)


In [41]:
%%timeit
walk_fast(1000)

35.3 µs ± 142 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


#### compare where is 35 using numpy and 761 using lists

In [53]:
np.arange(9).reshape(3,3).astype('int8')

array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]], dtype=int8)